build: GNU Make jobserver to prevent resource glut (make_job_server rebased on master)

jbohren opened the pull request:
This is a rebase and extension of @xqms's jobserver prototype in PR #85 which was created to fix #84. Without the jobserver, `catkin build` will use too many resources on machines with multiple cores.

Behavior using this PR:
- `catkin build --no-jobserver` -- current behavior (maxes out all jobs for all packages unless using another jobserver like distcc)
- `catkin build` -- build using the jobserver and as many jobs as CPUs
- `catkin build --jobserver` -- explicit version of the line above
- `catkin build -jN`-- build using the jobserver and `N` jobs
- `MAKEFLAGS="-jN" catkin build`-- build using the jobserver and `N` jobs
- `catkin config --no-jobserver` -- disable the jobserver for future builds
- `catkin build -lV` -- don't create more than one job at a time if the system load is greater than `V`
- `catkin build --mem-limit P` -- don't create more than one job at a time if more than `P`% of the system memory is used

Outstanding issues:
- [x] The jobserver behavior isn't documented
- [x] The status line reads `?/0 Jobs` when the jobserver is disabled, it should display something friendlier
- [x] change `internal_make_jobserver` switches `use_internal_make_jobserver`
 - https://github.com/catkin/catkin_tools/pull/155/files#r26704312
 - https://github.com/catkin/catkin_tools/pull/155/files#r26705185
- [x] [document `configure_make_args`](https://github.com/catkin/catkin_tools/pull/155/files#r26703936)
- [x] clean up jobserver API
 - [checking support](https://github.com/catkin/catkin_tools/pull/155/files#r26704509)
- [x] make `psutil` an optional dependency, and only load it if the memory control options are used

Future work:
- The number of jobs (`-jN`) and the number of packages (`-pM`) are still decoupled. The jobserver will limit the number of `Make` instances, but maybe this is ok?
- The current internal jobserver implementation does not monitor load or memory usage. These would be good arguments to support.

jbohren commented:
@xqms I've rebased this on master, it appears to work, and feel free to copy this snapshot to your fork.

jbohren commented:
@davetcoleman when you're feeling bored, give this branch a shot and see if you still run out of resources

jbohren commented:
Rebase of #85

davetcoleman commented:
I just pulled that branch and am testing now.

davetcoleman commented:
It kept my memory usage nice and stable - think whatever this is helped a lot!
![screenshot from 2015-03-09 22 57 56](https://cloud.githubusercontent.com/assets/561060/6569438/da6fe8fc-c6af-11e4-80e8-a49fc44d2280.png)

jbohren commented:
@xqms I modified the implementation so that it's possible to disable the jobserver with the `--ext-jobserver` argument. This way you can still use `distcc` like so:

```
CC="distcc gcc" CXX="distcc g++" catkin build -p$(distcc -j) -j$(distcc -j) --ext-jobserver
```

xqms commented:
@davetcoleman Nice to hear it helped you as well! `catkin_tools` is almost unusable without it on my workspace...

@jbohren okay, sounds good, though I think @wjwwood preferred an opt-in (like `--enable-jobserver`). I'll have time later today or tomorrow to take a look. If you guys want to go ahead changing/merging, don't wait for me ;-)

jbohren commented:
@xqms @wjwwood after seeing Dave's example workspace, I think that this should be the default behavior.

jbohren commented:
@xqms also is there a difference between `--jobserver-limit` and the normal `-j` jobs argument?

jbohren commented:
Something that might make sense is to add the jobserver options to the persistent config. This patch just augments the `--make-args` with the jobserver flags like `--jobserver-fds=3,4 -j` which makes it really visible, but also obfuscated.

Alternatively, it'd be better to have a high-level `Job Server: [internal / external]` option.

This would display like:

```
----------------------------------------------------------------------------
Profile:                     default
Extending:          [cached] /opt/ros/hydro
Workspace:                   /home/jbohren/ws/ascent
Source Space:       [exists] /home/jbohren/ws/ascent/src
Build Space:        [exists] /home/jbohren/ws/ascent/build
Devel Space:        [exists] /home/jbohren/ws/ascent/devel
Install Space:     [missing] /home/jbohren/ws/ascent/install
DESTDIR:                     None
----------------------------------------------------------------------------
Isolate Develspaces:         False
Install Packages:            False
Isolate Installs:            False
----------------------------------------------------------------------------
Additional CMake Args:       -DOBJREC_USE_CUDA=On -DOBJREC_CUDA_DEVICE_ID=1 
Additional Make Args:        None
Additional catkin Make Args: None
Make Job Server:             Internal
----------------------------------------------------------------------------
Workspace configuration appears valid.
----------------------------------------------------------------------------
```

xqms commented:
@jbohren `-j N` controls how many processes a `make` process may spawn. Since `catkin_tools` starts up to `-p X` parallel `make` processes this is useless for controlling the system load. `--jobserver-limit` specifies how many jobs (e.g. gcc processes) may be run in **total**, which is much more useful.

If the jobserver is enabled, it may make sense to replace `-j` and `-p` altogether with a single `-j` option controlling the jobserver limit...

Yes, I agree this thing should be integrated with the workspace config.

davetcoleman commented:
This is still running ok for me and catkin no longer crashes my computer, but it still uses a huge amount of memory sometimes while compiling a large workspace. Just now it got pretty close:

![screenshot from 2015-03-11 16 34 37](https://cloud.githubusercontent.com/assets/561060/6608450/b212f1f2-c80c-11e4-9a9f-7266e7568b8f.png)

jbohren commented:
Can you show the memory trace using catkin_make also?

wjwwood commented:
Thanks for everyone helping out on this one. I'll try to review it and try it out myself by the end of the week.

jack-oquin commented:
> If the jobserver is enabled, it may make sense to replace ``-j`` and ``-p`` altogether with a single ``-j`` option controlling the jobserver limit...

+1 That is what most of us naively *expected* ``-j`` to mean in the first place.

davetcoleman commented:
You mean the memory history? I'll try to record it next time I rebuild

jbohren commented:
Yeah.

On Thu, Mar 12, 2015, 16:58 Dave Coleman <notifications@github.com> wrote:

> You mean the memory history? I'll try to record it next time I rebuild
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/catkin/catkin_tools/pull/155#issuecomment-78614116>.
>

jbohren commented:
This still needs a bit of work to do what people "expect" it to do. Each job calls `handle_make_arguments` a couple times and I don't think that's where `MAKEFLAGS` should be read. It needs to be read early on and act as a default configuration. Then that needs to initialize the jobserver.

jbohren commented:
@xqms @davetcoleman @wjwwood This should be good to review. I re-worked it to make sure the jobserver singleton gets initialized in one place, but this could still be cleaned up.

The behavior of this PR is as follows:
- `catkin build --no-jobserver` -- current behavior (maxes out all jobs for all packages unless using another jobserver like distcc)
- `catkin build` -- build using the jobserver and as many jobs as CPUs
- `catkin build --jobserver` -- explicit version of the line above
- `catkin build -jN`-- build using the jobserver and `N` jobs
- `MAKEFLAGS="-jN" catkin build`-- build using the jobserver and `N` jobs

xqms commented on catkin_tools/argument_parsing.py:
```diff
@@ -167,6 +183,33 @@ def extract_cmake_and_make_arguments(args):
+    :rtype: dict
+    """
+
+    regex = r'(?:^|\s)(?:-?(j|l)(\s*[0-9]+|\s|$))' + \
```
nit: you probably wanted to use JOBS_FLAGS_REGEX from above here, right?

jbohren replied:
Ah yeah I meant to remove that in favor of just having two different patterns.

wjwwood replied:
Is this comment still unaddressed?

jbohren replied:
No, it's outdated. The referenced variable isn't in the PR any more.

xqms commented:
@jbohren I had a look at the code, seems good to me. I'll test this in my setup today and let you know if anything breaks.

xqms commented:
I've been running this for 2 days now with high compilation workloads in my setup, everything works as expected.

wjwwood commented:
lgtm, I'll use this while building up Indigo/Jade on my new laptop to try it out.

wjwwood commented on catkin_tools/argument_parsing.py:
```diff
@@ -201,32 +241,75 @@ def handle_make_arguments(input_make_args, force_single_threaded_when_running_te
+    :param internal_make_jobserver: if true, use the internal jobserver
+    :type make_args: bool
+    :rtype: tuple
+    """
```
The return tuple of this function is complicated, I think it would be really helpful for others if what is in it is spelled out in the `:returns:` directive or in the body of the doc string.

jbohren replied:
Yep.

wjwwood commented on catkin_tools/argument_parsing.py:
```diff
@@ -201,32 +241,75 @@ def handle_make_arguments(input_make_args, force_single_threaded_when_running_te
+
+    :param make_args: arguments to be passed to GNU Make
+    :type make_args: list
+    :param internal_make_jobserver: if true, use the internal jobserver
```
`use_internal_make_jobserver` would be a more clear variable name.

jbohren replied:
Agreed.

wjwwood commented on catkin_tools/argument_parsing.py:
```diff
@@ -201,32 +241,75 @@ def handle_make_arguments(input_make_args, force_single_threaded_when_running_te
+        internal_make_jobserver)
+
+    # If the jobserver is supported and enabled, override the jobs flags
+    using_jobserver = jobserver.supported and internal_make_jobserver
```
It's weird that you pass `internal_make_jobserver` to `get_instance` (and therefore the constructor of `MakeJobServer`), only to check `supported` AND `internal_make_jobserver` here. Is it not just sufficient to check `supported` here?

jbohren replied:
Yeah, there's no reason to check both, this is just slop. I'd actually like it if `supported` didn't keep having to be checked.

wjwwood commented on catkin_tools/context.py:
```diff
@@ -194,6 +196,8 @@ def __init__(
         :param make_args: extra make arguments to be passed to make for each package
         :type make_args: list
+        :param internal_make_jobserver: true if this configuration should use an internal make jobserv
+        :type internal_make_jobserver: bool
```
Again, throughout the code, `use_internal_make_jobsever` or something similar that indicates this is an option and not the actually job sever would be better.

wjwwood commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,151 @@
+
+from __future__ import print_function
+
+from os import pipe, write, close, unlink, read
```
Please do not do multiple imports on one line.

wjwwood commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,151 @@
+from os import pipe, write, close, unlink, read
+from multiprocessing import cpu_count
+from tempfile import mkstemp
+from subprocess import call, PIPE
```
Same here.

wjwwood commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,151 @@
+    This class implements a GNU make job server.
+
+    TODO:
+     - use os.getloadavg() to maintain server load
```
Is this TODO still relevant?

jbohren replied:
Yeah. Right now the jobserver only cares about the number of jobs, and not the system load average. It'd be nice if it could consider also the server load and the RAM footprint.

wjwwood replied:
Ok, let's just remember to open an issue to track this if we merge this without addressing that.

jbohren replied:
Yeah it's already mentioned in the "Future Work" section at the top of the PR.

wjwwood commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,151 @@
+                if e.errno == errno.EINTR:
+                    continue
+                else:
+                    raise
```
Simplify to:

```python
if e.errno != errno.EINTR:
    raise
```

wjwwood commented on catkin_tools/verbs/catkin_build/build.py:
```diff
@@ -732,12 +733,24 @@ def set_error_state(error_state):
-                        total_packages
-                    )
+
+                    if MakeJobServer.get_instance().supported:
```
I guess you're assuming here that `get_instance()` was already called, otherwise it will call the constructor with no args. Could that be a problem? Should there be a function to get and instance only if it already exists, otherwise fail?

wjwwood commented on catkin_tools/verbs/catkin_build/build.py:
```diff
@@ -732,12 +733,24 @@ def set_error_state(error_state):
+                    if MakeJobServer.get_instance().supported:
+                        msg_rhs = clr("[{0}/{1} Jobs][{2}/{3} Packages][{4}/{5} Completed]").format(
+                            MakeJobServer.get_instance().num_running_jobs(),
+                            MakeJobServer.get_instance().num_jobs,
```
Is there any reason to continue to get the instance? Are you expecting it to asynchronously change? If not it would be better to store it and reuse it to reduce function calls (which do cost in Python).

wjwwood commented on catkin_tools/verbs/catkin_build/build.py:
```diff
@@ -732,12 +733,24 @@ def set_error_state(error_state):
-                    )
+
+                    if MakeJobServer.get_instance().supported:
+                        msg_rhs = clr("[{0}/{1} Jobs][{2}/{3} Packages][{4}/{5} Completed]").format(
```
This breaks the colorization of the part of the status message. You'll need to provide a matching entry in this file:

https://github.com/catkin/catkin_tools/blob/0afc6f9c21c933348e48a3b3f2da1da63d2a4865/catkin_tools/verbs/catkin_build/color.py#L57-L58

jbohren commented:
@wjwwood I've cleaned up the API a bunch based on the discussion before

I also added the following
- basic load average monitoring
- an experimental memory usage monitoring

All rebased on latest master.

NikolausDemmel commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,233 @@
+            # make sure we're observing load maximums
+            if self.max_load is not None:
+                try:
+                    max_load = 8.0
```
This line is probably left over by accident?

NikolausDemmel commented on catkin_tools/make_jobserver.py:
```diff
@@ -0,0 +1,233 @@
+                if jobserver_running_jobs() > 0 and mem_usage.percent > self.max_mem:
+                    time.sleep(0.01)
+                    continue
+
```
If both `self.max_load` and `self.max_mem` are set, you probably want to wait for both conditions in a single loop, right?

jbohren replied:
Yeah, it's already in a single loop.

NikolausDemmel replied:
My bad. My parsing was off. Sorry for the noise.

xqms commented:
I have just discovered an issue: Multiple invocations of `catkin build --save` duplicate the `--jobserver-fds=3,4 -j` arguments. Looks like this:

```
----------------------------------------------
Additional CMake Args:       -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS=-Wall
Additional Make Args:        --jobserver-fds=3,4 -j --jobserver-fds=3,5 -j
Additional catkin Make Args: None
Internal Make Job Server:    True
----------------------------------------------
```

xqms commented:
Okay, to clarify: the `--jobserver-fds` argument is saved in the workspace config. It shouldn't be, since it is also generated by the "Internal Make Job Server" flag.

jbohren commented:
@xqms Yeah, also now that `Internal Make Job Server:    True` is in the summary, I don't think the jobserver flags should be added in the way that I have them added. I'll make it so it displays the summary like this:

```
----------------------------------------------
Additional CMake Args:       -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_CXX_FLAGS=-Wall
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True, Args: "--jobserver-fds=3,4 -j"
----------------------------------------------
```

wjwwood commented:
https://travis-ci.org/catkin/catkin_tools/jobs/56253398#L1432

jbohren commented:
@wjwwood Just updated. There was an issue previously where the make arg parsing wasn't properly extracting jobs args formatted like `-j N` with a space between the `-j` and the `N`.

@xqms This removes the jobserver args from "additional make args". I was considering adding the jobserver max jobs / max load / max mem to the context, but we'd also have to add those options to the saved configs, which just involves a lot of additional changes.

As it stands, I think that this is a pretty critical patch and if there are any other outstanding features, we should bump them to another PR.

It would be nice to add a strong test harness around this thing, but I'm not sure what the best approach would be for that. Mostly you'd want to test:

 - defaults
 - cli args
 - MAKEFLAGS args
 - disabling the jobserver

davetcoleman commented:
I've been using, and keeping up with the updates, of this branch everyday. Seems to work for me.

wjwwood commented:
You've not added the dependency on `psutil` correctly, you also have to update some other files:

- https://github.com/catkin/catkin_tools/blob/master/setup.py
- https://github.com/catkin/catkin_tools/blob/master/stdeb.cfg

Also the function you're using from `psutil` is deprecated:

https://code.google.com/p/psutil/wiki/Documentation#Memory

Please at least update the pull request to use `psutil.virtual_memory()`.

I'm not happy about adding a dependency on `psutil`, but if there isn't a reasonable way to do it without `psutil` then I guess we'll have to. One option would be to not import psutil unless a user tries to use the memory limits.

Having said all of that, I'm really not convinced the memory limits is going to be a reliable way to manage how much memory the whole compile is using. Unlike CPU usage, the memory usage of each `cc` in a make job can grow over time. Imagine:

- each of 6 `cc` process will approach 2 GB of memory usage over 30 seconds (this behavior is common with some of the `pcl` object files)
- your system has 4 GB of RAM
- you compiled with -j6
- and you've set the memory limit to 90%

You'll hand out all six job tokens (because for the first few seconds the memory usage is really low) and each of them will grow to use 2 GB and run your system out of memory anyways. However, it's possible that it won't work out like that and the memory limit will help, but I wouldn't call it reliable.

This is a tough problem and we worked with the Tango project to work around this problem in a few places. The most reliable way to fix this kind of "perfect storm of parallel make jobs" was to instrument the build so we could figure out which process, when run at the same time, busted the system. Then they would manually adjust the inter-target dependencies to prevent those object files from being built at the same time. @tfoote was the instrumenting code ever pushed somewhere public?

Based on that, I would say we should either remove the memory limit option or make it "experimental" and make the dependency on `psutil` optional. If a user is savvy enough to make use of the memory limit options then they are probably also ok with installing `psutil` when prompted.

jbohren commented:
> Having said all of that, I'm really not convinced the memory limits is going to be a reliable way to manage how much memory the whole compile is using. Unlike CPU usage, the memory usage of each cc in a make job can grow over time. Imagine:

Of course it's not going to be reliable. But if some crazy person (@davetcoleman *cough* *cough*) is running around with a computer with no swap, he can probably set it to something like 65% and be reasonably confident it's not going to brick his machine.  

> This is a tough problem and we worked with the Tango project to work around this problem in a few places. The most reliable way to fix this kind of "perfect storm of parallel make jobs" was to instrument the build so we could figure out which process, when run at the same time, busted the system. Then they would manually adjust the inter-target dependencies to prevent those object files from being built at the same time. @tfoote was the instrumenting code ever pushed somewhere public?

"Perfect is the enemy of the good" -- *some guy that got something done*

I'm not concerned enough about this problem to do any more about it than is in the experimental memory limit check.

> Based on that, I would say we should either remove the memory limit option or make it "experimental" and make the dependency on psutil optional. If a user is savvy enough to make use of the memory limit options then they are probably also ok with installing psutil when prompted.

It already is an unlisted "experimental" option. I'm happy putting a try-except around the psutil import and then throwing an error if someone tries using it without support. That being said, anyone with `rqt` installed will have `python-psutil`.

> You've not added the dependency on psutil correctly, you also have to update some other files:

> - https://github.com/catkin/catkin_tools/blob/master/setup.py
> - https://github.com/catkin/catkin_tools/blob/master/stdeb.cfg

I'm going to remove the dependency description completely (except for the travis config) because it's going to be an optional dependency.

>Also the function you're using from psutil is deprecated:

> https://code.google.com/p/psutil/wiki/Documentation#Memory

> Please at least update the pull request to use psutil.virtual_memory().

Yeah I know it's deprecated. Unfortunately the `python-psutil` package on 12.04 doesn't have this function yet. Also ROS Hydro depends on the `python-psutil` package, so ditching it in favor of the pypi doesn't sound like a good plan. It wouldn't be too hard to just check if the function is available and use the new one if the old one isn't available.

wjwwood commented:
> It already is an unlisted "experimental" option. I'm happy putting a try-except around the psutil import and then throwing an error if someone tries using it without support. That being said, anyone with rqt installed will have python-psutil.

I would appreciate making it optional. It's not going to be a problem for ROS users, but I put a lot of effort into minimizing the dependencies, so I'd like to keep it that way if possible.

> Yeah I know it's deprecated. Unfortunately the python-psutil package on 12.04 doesn't have this function yet. Also ROS Hydro depends on the python-psutil package, so ditching it in favor of the pypi doesn't sound like a good plan. It wouldn't be too hard to just check if the function is available and use the new one if the old one isn't available.

I remember this now, it was quite a problem for `rqt_top`. Unfortunately the `pypi` version is never an option because we cannot build a `.deb` that depends on something in `pypi`.

If you can get the optional import in place tonight or tomorrow morning I can merge this and get it in the next release. Tomorrow is our ROS bug fix party, so I plan to spend some time on this tracker and making a release.

Thanks for working on it and iterating with me.

tfoote commented:
@wjwwood re: collecting build parameters of specific invocations. The tools is available at https://github.com/osrf/watchprocess  It can generate a report of cpu and memory usage for each process using an instrumented program during a build.

davetcoleman commented:
Thanks for the crazy person shout out @jbohren I've taken your advice:
![screenshot from 2015-04-02 12 07 20](https://cloud.githubusercontent.com/assets/561060/6970162/f3d70b5c-d930-11e4-86cc-33963a44caf8.png)

jbohren commented:
@wjwwood I added the lazy import of psutil and the version checking. I've tested it with version 0.4.1 but no other versions.

wjwwood commented:
I merged this with some adjustments.

wjwwood commented:
Also, a rebase which is why it doesn't show up as "merged".
